signer/kms/aws/signer.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/in-toto/go-witness/log"
+	kms "github.com/in-toto/go-witness/signer/kms"
+)
+
+var awsSupportedAlgorithms = []types.CustomerMasterKeySpec{
+	types.CustomerMasterKeySpecRsa2048,
+	types.CustomerMasterKeySpecRsa3072,
+	types.CustomerMasterKeySpecRsa4096,
+	types.CustomerMasterKeySpecEccNistP256,
+	types.CustomerMasterKeySpecEccNistP384,
+	types.CustomerMasterKeySpecEccNistP521,
+}
+
+var awsSupportedHashFuncs = []crypto.Hash{
+	crypto.SHA256,
+	crypto.SHA384,
+	crypto.SHA512,
+}
+
+// SignerVerifier is a cryptoutil.SignerVerifier that uses the AWS Key Management Service
+type SignerVerifier struct {
+	reference string
+	client    client
+	hashFunc  crypto.Hash
+}
+
+// LoadSignerVerifier generates signatures using the specified key object in AWS KMS and hash algorithm.
+func LoadSignerVerifier(ctx context.Context, ksp *kms.KMSSignerProvider) (*SignerVerifier, error) {
+	a := &SignerVerifier{
+		reference: ksp.Reference,
+	}
+
+	var err error
+	a.client, err = newAWSClient(ctx, ksp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, hashFunc := range awsSupportedHashFuncs {
+		if hashFunc == ksp.HashFunc {
+			a.hashFunc = ksp.HashFunc
+		}
+	}
+
+	if a.hashFunc == 0 {
+		return nil, fmt.Errorf("unsupported hash function: %v", ksp.HashFunc)
+	}
+
+	return a, nil
+}
+
+// NOTE: This might be all wrong but setting it like so for now
+//
+// KeyID returns the key identifier for the key used by this signer.
+func (a *SignerVerifier) KeyID() (string, error) {
+	return a.reference, nil
+}
+
+// Sign signs the provided message using AWS KMS. If the message is provided,
+// this method will compute the digest according to the hash function specified
+// when the Signer was created.
+func (a *SignerVerifier) Sign(message io.Reader) ([]byte, error) {
+	var err error
+	ctx := context.Background()
+	var digest []byte
+
+	var signerOpts crypto.SignerOpts
+	signerOpts, err = a.client.getHashFunc(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting fetching default hash function: %w", err)
+	}
+
+	hf := signerOpts.HashFunc()
+
+	digest, _, err = cryptoutil.ComputeDigest(message, hf, awsSupportedHashFuncs)
+	if err != nil {
+		return nil, err
+	}
+
+	return a.client.sign(ctx, digest, hf)
+}
+
+// Verifier returns a cryptoutil.Verifier that can be used to verify signatures created by this signer.
+func (a *SignerVerifier) Verifier() (cryptoutil.Verifier, error) {
+	return a, nil
+}
+
+// Bytes returns the bytes of the public key that can be used to verify signatures created by the signer.
+func (a *SignerVerifier) Bytes() ([]byte, error) {
+	ctx := context.Background()
+	p, err := a.client.fetchPublicKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return cryptoutil.PublicPemBytes(p)
+}
+
+// Verify verifies the signature for the given message, returning
+// nil if the verification succeeded, and an error message otherwise.
+func (a *SignerVerifier) Verify(message io.Reader, sig []byte) (err error) {
+	ctx := context.Background()
```

Comment:
I think long term we'll want to change the interface to allow for passing the context in directly. Contexts can be tied to specific requests and used to cancel requests to external resources -- which we want here.

We can punt on it for now, but I'd recommend changing these `context.Background` calls to `context.TODO()` instead for the time being

Reply:
Yeah I'lll set this as a `TODO()` for myself!